src/chainstate/stacks/db/blocks.rs

```diff
@@ -3380,152 +3396,247 @@ impl StacksChainState {
+        let cnt = query_count(
+            blocks_conn,
+            &sql,
+            &[&u64_to_sql(min_arrival_time)?, &u64_to_sql(limit)?],
+        )
+        .map_err(Error::DBError)?;
+        Ok(cnt as u64)
+    }
+
+    /// Measure how long a block waited in-between when it arrived and when it got processed.
+    /// Includes both orphaned and accepted blocks.
+    pub fn measure_block_wait_time(
+        blocks_conn: &DBConn,
+        start_height: u64,
+        end_height: u64,
+    ) -> Result<Vec<i64>, Error> {
+        let sql = "SELECT processed_time - arrival_time FROM staging_blocks WHERE processed = 1 AND height >= ?1 AND height < ?2";
+        let args: &[&dyn ToSql] = &[&u64_to_sql(start_height)?, &u64_to_sql(end_height)?];
+        let list = query_rows::<i64, _>(blocks_conn, &sql, args)?;
+        Ok(list)
+    }
+
+    /// Measure how long a block took to be downloaded (for blocks that we downloaded).
+    /// Includes _all_ blocks.
+    pub fn measure_block_download_time(
+        blocks_conn: &DBConn,
+        start_height: u64,
+        end_height: u64,
+    ) -> Result<Vec<i64>, Error> {
+        let sql = "SELECT download_time FROM staging_blocks WHERE height >= ?1 AND height < ?2";
+        let args: &[&dyn ToSql] = &[&u64_to_sql(start_height)?, &u64_to_sql(end_height)?];
+        let list = query_rows::<i64, _>(blocks_conn, &sql, args)?;
+        Ok(list)
     }
 
     /// Given access to the chain state (headers) and the staging blocks, find a staging block we
     /// can process, as well as its parent microblocks that it confirms
     /// Returns Some(microblocks, staging block) if we found a sequence of blocks to process.
     /// Returns None if not.
-    fn find_next_staging_block(
-        blocks_conn: &DBConn,
+    fn find_next_staging_block<'a>(
+        blocks_tx: &mut BlocksDBTx<'a>,
         blocks_path: &String,
         headers_conn: &DBConn,
+        sort_conn: &DBConn,
     ) -> Result<Option<(Vec<StacksMicroblock>, StagingBlock)>, Error> {
         test_debug!("Find next staging block");
 
-        // go through staging blocks and see if any of them match headers and are attachable.
-        // pick randomly -- don't allow the network sender to choose the processing order!
-        let sql = "SELECT * FROM staging_blocks WHERE processed = 0 AND attachable = 1 AND orphaned = 0 ORDER BY RANDOM()".to_string();
-
-        let mut stmt = blocks_conn
-            .prepare(&sql)
-            .map_err(|e| Error::DBError(db_error::SqliteError(e)))?;
+        let mut to_delete = vec![];
 
-        let mut rows = stmt
-            .query(NO_PARAMS)
-            .map_err(|e| Error::DBError(db_error::SqliteError(e)))?;
+        // put this in a block so stmt goes out of scope before we start to delete PoX-orphaned
+        // blocks
+        {
+            // go through staging blocks and see if any of them match headers, are attachable, and are
+            // recent (i.e. less than 10 minutes old)
```

Comment:
Where is this recency enforced? I don't see a recency check in `can_attach`. Also, why do we want to only attempt to attach recently received blocks?

Reply:
The comment is wrong -- there is no recency check.  I was experimenting with biasing block selection by recency at one point, but that code no longer exists.